Source/DafnyRuntime/DafnyRuntime.py

```diff
@@ -18,17 +18,25 @@ def print(value):
 class Break(Exception):
     target: str
 
+class TailCall(Exception):
+    pass
+
 @contextmanager
-def label(name: str):
+def label(name: str = None):
     try:
         yield
     except Break as g:
         if g.target != name:
             raise g
+    except TailCall:
+        pass
```

Comment:
This is slightly wrong, because if you have a tail-recursive function with a label inside, that label will catch the `TailCall` exception at the wrong spot. We should add a test case like that. :)

Reply:
Sorry, but I don't follow. Can you give a concrete example?

Reply:
Here's a test case for what I'm worried about:

```dafny
method {:tailrecursion true} Countdown(x: nat) {
  label here: {
    if x == 0 {
      print "Kablammo!!\n";
    } else {
      print x, "\n";
      Countdown(x - 1);
    }
  }
}

method Main() {
  Countdown(5);
}
```

```
$ dafny /compile:3 /spillTargetCode:1 /compileTarget:cs src/Scratch.dfy 

Dafny program verifier finished with 2 verified, 0 errors
Wrote textual form of target program to Scratch.cs
Running...

5
4
3
2
1
Kablammo!!
$ dafny /compile:3 /spillTargetCode:1 /compileTarget:py src/Scratch.dfy 

Dafny program verifier finished with 2 verified, 0 errors
Wrote textual form of target program to Scratch-py/Scratch.py
Additional target code written to Scratch-py/_dafny.py
Additional target code written to Scratch-py/System_.py
Additional target code written to Scratch-py/module_.py
Running...

5 
```

The issue is that the `label here: { ... }` statement catches the `TailCall` exception instead of the label emitted by `EmitTailCallStructure`, so the tail recursion never happens. The source looks like this:

```python
class default__:
    @staticmethod
    def Countdown(x):
        while True:
            with _dafny.label():
                with _dafny.label("0"):
                    if True:
                        if (x) == (0):
                            _dafny.print("Kablammo!!\n")
                        elif True:
                            _dafny.print(x)
                            _dafny.print("\n")
                            in0_ = (x) - (1)
                            x = in0_
                            _dafny._tail_call()
                break
```

We just need some way for the `label()` handler to know when it is meant to catch `TailCall` vs. `Break`.